### DISCUSSION TITLE
Notification Applications Not Quitting?

### DISCUSSION THREAD

**Discussion by Oh4sh0** [IBM/mac-ibm-notifications#47] [Q&A] [answered]
Hello,

I've been testing the tool without rebuilding it for my own usage yet, on 11.6, and I've found that when the binary is invoked, even with a timeout value, it seems to hang and never exit. 

Example:
/Users/xxx/Desktop/IBM\ Notifier\ 2021-09-23\ 08-05-35/IBM\ Notifier.app/Contents/MacOS/IBM\ Notifier -title "Test title" -subtitle "Test subtitle" -payload "text" -timeout 5

The process just sits in activity monitor thereafter running in the background. I haven't seen other reports of this behavior so I figured it must be something unique to me, but I can reproduce it across multiple devices. 

Ex:
![IBM_Notifier](https://user-images.githubusercontent.com/62770756/134511611-1630708a-ae0e-428e-8b52-5d5599e31ba5.png)

Cheers,
Casey

**Comment by SMartorelli** [comment]
Hi @Oh4sh0, the agent core and every component should exit with a return value or an error so it should not hang in background. In the example you posted there are some errors:

1. `-type` argument missing (Ex. `-type banner -title "This is a title"`)
2. `-payload` argument must be used only to define onboarding UI type structure (Ex. `-type onboarding -payload "JSON STRING"`)
3. `-timeout` argument must be used only for popup UI type (Ex. `-type popup -title "This is a title" -timeout 5`)

Please refer to our [Wiki](https://github.com/IBM/mac-ibm-notifications/wiki) for any doubt about the [Usage](https://github.com/IBM/mac-ibm-notifications/wiki/Usage) of **IBM Notifier**

Make sure to close all the background executions for `IBM Notifier *` and test it again. If the issue persist please post here the argument set that caused it. 

PS: We suggest you to use the latest version of IBM Notifier because it include all the latest fix and features. We're working to provide a signed&notarized version also for the just released 2.3.2 version.

**Comment by Oh4sh0** [comment]
Thank you @SMartorelli. 

This syntax would be correct then, at its simplest?
/Users/xxx/Desktop/IBM\ Notifier\ 2021-09-23\ 08-05-35/IBM\ Notifier.app/Contents/MacOS/IBM\ Notifier -type banner -title "This is a title" ?

With all instances closed, IBM Notifier and IBM Notifier Banner launch and remain running, indefinitely, unless I perform a "CTRL+C" to terminate the execution attempt.

**Comment by SMartorelli** [comment]
Yes it is correct, this behaviour happens on purpose. In this way you can track the user response to the banner (script exit code). Be aware that for banner UI type "almost" every action of the user is tracked but could happen that if the user close multiple banners together (from the Notification Center) this action couldn't be recognised by the agent and this is the only behaviour that cause the banner component to hang in background without any purpose. Because of this we suggest to use banners (temporary banners) for actions where the user answer is not mandatory (use alert or popup instead).
By the way the banner or the core component hanging out in background doesn't interfere with other executions of IBM Notifier and they're really lightweight as you can see.

**Comment by Oh4sh0** [comment]
If this were called directly in a policy/thread, eg via JAMF policy invoking the binary, wouldn't it tie up the thread because the execution may never be returned unless the user acknowledge/does something with the banner?

You note temporary banners, is there something making them temporary, because what I'm seeing is that again, that execution/process persists unless the user acknowledges/clicks on the banner. Is there an integrated timeout?

**Reply by SMartorelli** [reply]
Well it's your choice to wait for the user response in the script, if you don't care about the user interaction you can send the execution of IBM Notifier to a subshell in order to not block your script on it.

I said temporary because banners hide their self in the notification center automatically after 5 seconds, alerts instead remain always visible.

**Reply by Oh4sh0** [reply]
But those processes would remain running indefinitely until the user reboots if they just leave them in the notification center?

I agree, there are other ways to deal with that behavior, just surprised it is the default.

**Reply by SMartorelli** [reply]
Yeah this could happen with banners. You can anyway quit the execution from the script, this will quit the two processes without making the banner disappear. If in your opinion this is something important to improve please we're more then welcome to receive a [Feature Request](https://github.com/IBM/mac-ibm-notifications/issues/new?assignees=&labels=&template=feature_request.md&title=) 😁

**Reply by Oh4sh0** [reply]
Thanks for the quick responses, Simone. I really do appreciate the interaction and guidance.

**Reply by Oh4sh0** [reply]
I suppose I'll continue to pick your brain here. Are y'all spawn a subprocess, etc? for this. When I attempt to do that with nohup to leave it running in the background, the application rockets in cpu usage. 
`nohup /Users/xxx/Desktop/IBM\ Notifier\ 2021-09-23\ 08-05-35/IBM\ Notifier.app/Contents/MacOS/IBM\ Notifier -type alert -title "This is a title" -main_button_label "Go" -main_button_cta_type link -main_button_cta_payload "https://www.ibm.com" > foo.out 2> foo.err < /dev/null &`

results in IBM Notifier sitting at 100% cpu usage.

**Reply by SMartorelli** [reply]
Uhm I see some strange behaviours here doing some tests...I need to figure it out where is the bug here tomorrow morning and I'll keep you posted. I'll might create an Issue starting from this discussion in case.

**Reply by Oh4sh0** [reply]
Alrighty. Thank you!

**Comment by SMartorelli** [comment, marked as answer]
Just created a new [issue](https://github.com/IBM/mac-ibm-notifications/issues/48) where you can track the fix to this unexpected behaviour @Oh4sh0
